### ISSUE TITLE
support __new__() in extension types

### ISSUE THREAD

**Issue by robertwb** [cython/cython#799]
Extension types should support the `__new__()` special method and be able to return any kind of object. The proposed syntax is described in #793 (migrated from ticket http://trac.cython.org/ticket/232):

```
cdef TYPE __new__(self, foo, bar)
```

Migrated from http://trac.cython.org/ticket/238

**Comment by oleksandr-pavlyk** [comment]
I would like to upvote this request. 

Attempting to build a subclass of `numpy.ndarray` object in Cython which implements `__new__` method and `__array_finilize__` method I encountered

```
__new__ method of extension type will change semantics in a future version of Pyrex and Cython. Use __cinit__ instead.
```

The message has been added about 10 years ago:

https://github.com/cython/cython/blob/fcfd16c7467c31f255287a73f36cf66b32bc096c/Cython/Compiler/Symtab.py#L2291

**Comment by oleksandr-pavlyk** [comment]
@scoder I could not figure out from referenced links whether this is considered as fixed, or is still an open issue.

Having Cython support `__new__` would allow writing subclasses of `numpy.ndarray` in Cython.

**Comment by scoder** [comment]
It's an open issue. AFAIK, no-one has really worked on it during the last 11 years. Contributions are welcome.

**Comment by da-woods** [comment]
> Having Cython support `__new__` would allow writing subclasses of `numpy.ndarray` in Cython.

@oleksandr-pavlyk You might be able to use two levels of inheritance - a `cdef class` and a regular class - with `__new__` defined in the regular class:

```
from numpy cimport ndarray

cdef class _C(ndarray):
    # implementation goes here

class C(_C):
    def __new__(self, *args, **kwds):
         # etc...

```

Provided you don't need to change the arguments it's possible you might not even need to do that. `__cinit__` ends up in the `tp_new` slot and automatically calls `ndarray.__new__` just forwarding the arguments it was passed. That might do for a lot of cases.

**Comment by J08nY** [comment]
Hey all, does anyone have any pointers on what is the blocker here? If just the check that errors out is removed will something reasonable happen?

Resolving this would really benefit me and I believe some other projects as well.

**Comment by da-woods** [comment]
I don't think it's simple. We currently put `__cinit__` in the `tp_new` slot, so it'd need careful thought about how to interact with that. 

So it's not just a case of "implementing something" - it needs a bit of design too.

**Comment by J08nY** [comment]
> So it's not just a case of "implementing something" - it needs a bit of design too.

Yeah, I played with it a bit and noticed that too. There seem to be two core issues:
 - `__cinit__` in the `tp_new`. Interplay between `__cinit__` and `__new__` would have to be defined somehow. What happens when both are defined, etc.?
 -  There is also a mention of the __new__ return type here: https://github.com/cython/cython/blob/746627fdca5e48db2b63c8bb4be82b33ea617549/Cython/Compiler/ExprNodes.py#L5953-L5958
   With support for `__new__` that can return an arbitrary subclass this would need to be resolved I guess.

**Comment by scoder** [comment]
> We currently put `__cinit__` in the `tp_new` slot

We don't. We put a function there that invokes the (`tp_new`) allocation function of the base class and then calls its own `__cinit__` on the object when coming back. That can still be done with a custom `__new__`, but it would mean that if `__new__` returns an object of a different class, there would be a different allocation chain with (possibly) their own entirely different `__cinit__` functions to be called that would be part of the allocation chain *invoked by* `__new__`, not the chain that *called* `__new__`.

There are different use cases to be looked at here, mainly:

Return an unrelated object, e.g.
```python
class C:
    def __new__(cls):
        return {}
```
There's nothing to do in this case, just invoke the user `__new__` method from `tp_new` and return the resulting object.

We can also return a related object from the hierarchy, e.g.
```python
class _Base:
    def __new__(cls):
        c = create_new_instance_of(cls)
        c.initialised = True
        return c

class Subclass(_Base):
    …

s = Subclass()
```
And this is where it becomes tricky. How do you do `create_new_instance_of(cls)`? Is that `object.__new__(cls)`? That can't be implemented as `object.tp_new(cls)` because that wouldn't be sufficient to initialise the object. It would still need to call `Subclass.tp_new()` and the entire call chain, but bypass `Base.__new__()`. So, `__new__` is actually not `tp_new`, semantically. `tp_new` is what should be called in the very end, by `object.__new__(cls)`, including the complete call chain of the class and all of its `__cinit__` methods.

I think it becomes easier to understand if it we consider `__new__` and `tp_new` unrelated features.
